src/Test.FunctionalTests.Sarif/Partitioning/PartitionTests.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using FluentAssertions;
+using Microsoft.CodeAnalysis.Sarif;
+using Microsoft.CodeAnalysis.Sarif.Visitors;
+using Microsoft.CodeAnalysis.Sarif.Writers;
+using Newtonsoft.Json;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.CodeAnalysis.Test.FunctionalTests.Sarif.Partitioning
+{
+    public class PartitionTests : FileDiffingFunctionalTests
+    {
+        public PartitionTests(ITestOutputHelper outputHelper, bool testProducesSarifCurrentVersion = true) :
+            base(outputHelper, testProducesSarifCurrentVersion)
+        { }
+
+        protected override string IntermediateTestFolder => @"Partitioning";
+
+        private class TestParameters
+        {
+            internal PartitioningVisitor<string>.PartitionFunction PartitionFunction { get; set; }
+            internal bool DeepClone { get; set; }
+        }
+
+        [Fact]
+        public void Partition_WithTrivialPartitionFunction_WithDeepClone_ReturnsLogWithAllResultsAndRunLevelArrayContentsFromAllResults()
+        {
+            Partition_WithTrivialPartitionFunction_ReturnsLogWithAllResultsAndRunLevelArrayContentsFromAllResults(deepClone: true);
+        }
+
+        [Fact]
+        public void Partition_WithTrivialPartitionFunction_WithShallowCopy_ReturnsLogWithAllResultsAndRunLevelArrayContentsFromAllResults()
+        {
+            Partition_WithTrivialPartitionFunction_ReturnsLogWithAllResultsAndRunLevelArrayContentsFromAllResults(deepClone: false);
+        }
+        private void Partition_WithTrivialPartitionFunction_ReturnsLogWithAllResultsAndRunLevelArrayContentsFromAllResults(bool deepClone)
+        {
+            PartitioningVisitor<string>.PartitionFunction partitionFunction = result => "default";
+
+            RunTest(
+                inputResourceNames: new List<string> { "Partition.sarif" },
+                expectedOutputResourceNames: new Dictionary<string, string>
+                {
+                    ["default"] = "TrivialPartitionFunction.sarif"
+                },
+                parameter: new TestParameters
+                {
+                    PartitionFunction = partitionFunction,
+                    DeepClone = deepClone
+                });
+        }
+
+        [Fact]
+        public void Partition_ByRuleId_WithDeepClone_ProducesOneLogFilePerRule()
+        {
+            Partition_ByRuleId_ProducesOneLogFilePerRule(deepClone: true);
+        }
+
+        [Fact]
+        public void Partition_ByRuleId_WithShallowCopy_ProducesOneLogFilePerRule()
+        {
+            Partition_ByRuleId_ProducesOneLogFilePerRule(deepClone: false);
+        }
+
```

Comment:
If you have a sample with more than one run, you should check that Runs are excluded altogether if none of the Run Results is in a given partition. (If not already covered by the above)

Reply:
The test cases do already cover that. The input file Partition.sarif has two runs. Note that the output file TST0002.sarif contains only one run (corresponding to the second of the two runs in the input file), because only that run has any results with `ruleId` `TST0002`.

---
In reply to: [357432826](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357432826) [](ancestors = 357432826)